{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14492e62-ee7f-47d7-bca9-771d6a54b0a1",
   "metadata": {},
   "source": [
    "# Identification of Potential Drug Targets of Inflammatory Bowel Disease\n",
    "\n",
    "*Please note: This notebook uses open access data*  \n",
    "\n",
    "#### Fan Wang\n",
    "#### July 31 2022\n",
    "\n",
    "Systematic MR of molecular phenotypes such as proteins and expression of transcript levels offer enormous potential to prioritise drug targets for further investigation. However, many genes and gene products are not easily druggable, so some potentially important causal genes may not offer an obvious route to intervention.\n",
    "\n",
    "A parallel problem is that current GWAS of molecular phenotypes have limited sample sizes and limited protein coverages. A potential way to address both these problems is to use protein-protein interaction (PPI) information to identify druggable targets which are linked to a non-druggable, but robustly causal target. Their relationship to the causal target increases our confidence in their potential causal role even if the initial evidence of effect is below our multiple-testing threshold.\n",
    "\n",
    "This notebook demonstrates an approach to query data in [EpiGraphDB](https://api.epigraphdb.org/) to prioritize potential alternative drug targets in the same PPI network for Inflammatory Bowel Disease (IBD), as follows:\n",
    "\n",
    "* For an existing drug target of interests, we use PPI networks to search for its directly interacting genes that are evidenced to be druggable.\n",
    "* We then examine the causal evidence of these candidate genes on the disease.\n",
    "* We also examine the literature evidence of these candidate genes on the disease.\n",
    "* Then we query the metadata including meta nodes and meta edges and the overall schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1bc9c780-9011-4d36-8ac0-355aa2816552",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from pprint import pprint\n",
    "from pprint import pformat\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "get_ipython().run_line_magic(\"config\", \"InlineBackend.figure_format = 'svg'\")\n",
    "import networkx as nx\n",
    "import requests\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2043278f-aea8-4fe9-b750-3b67155a23df",
   "metadata": {},
   "source": [
    "First, we will ping the API to check our connection. Here we use the .get() method to send a GET request to the `/ping` endpoint of the API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "facde7fa-491e-4789-bd72-7e097e59391a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "If this line gets printed, ping was sucessful.\n"
     ]
    }
   ],
   "source": [
    "API_URL = \"https://api.epigraphdb.org\"\n",
    "endpoint = \"/ping\"\n",
    "response_object = requests.get(API_URL + endpoint)\n",
    "GENE_NAME = \"IL23R\"\n",
    "OUTCOME_TRAIT = \"Inflammatory bowel disease\"\n",
    "\n",
    "# Check that the ping was sucessful\n",
    "try:\n",
    "    response_object.raise_for_status()\n",
    "    print(\"If this line gets printed, ping was sucessful.\")\n",
    "except requests.exceptions.HTTPError as err:\n",
    "    print(err)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae00cc0c-18c3-42fe-ba97-a48b27bd79fb",
   "metadata": {},
   "source": [
    "### 1. Using PPI networks for alternative drug targets search\n",
    "The assumption here is that the most likely alternative targets are either directly interacting with IL23R or somewhere in the PPI network. In this example, we consider only genes that were found to interact with *IL23R* via direct protein-protein interactions, and require that those interacting proteins should also be druggable.\n",
    "\n",
    "The thousands of genes are classified with regard to their druggability by Finan et al. 2017, where the Tier 1 category refers to approved drugs or those in clinical testing while for other tier categories the druggability confidence drops in order Tier 2 and then Tier 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3bc85d77-551a-4a58-9271-5758e74ed2f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1.name</th>\n",
       "      <th>p1.uniprot_id</th>\n",
       "      <th>p2.uniprot_id</th>\n",
       "      <th>g2.name</th>\n",
       "      <th>g2.druggability_tier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P04141</td>\n",
       "      <td>CSF2</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P01562</td>\n",
       "      <td>IFNA1</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P01579</td>\n",
       "      <td>IFNG</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P22301</td>\n",
       "      <td>IL10</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P29460</td>\n",
       "      <td>IL12B</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P42701</td>\n",
       "      <td>IL12RB1</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P35225</td>\n",
       "      <td>IL13</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P40933</td>\n",
       "      <td>IL15</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q16552</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q96PD4</td>\n",
       "      <td>IL17F</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P60568</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9GZX6</td>\n",
       "      <td>IL22</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9NPF7</td>\n",
       "      <td>IL23A</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P05112</td>\n",
       "      <td>IL4</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P05113</td>\n",
       "      <td>IL5</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P05231</td>\n",
       "      <td>IL6</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P15248</td>\n",
       "      <td>IL9</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P23458</td>\n",
       "      <td>JAK1</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>O60674</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P19838</td>\n",
       "      <td>NFKB1</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P42336</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P51449</td>\n",
       "      <td>RORC</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P40763</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q969D9</td>\n",
       "      <td>TSLP</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P29597</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>Tier 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P51684</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P25963</td>\n",
       "      <td>NFKBIA</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9HC29</td>\n",
       "      <td>NOD2</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P27986</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q04206</td>\n",
       "      <td>RELA</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P42224</td>\n",
       "      <td>STAT1</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P42229</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P42226</td>\n",
       "      <td>STAT6</td>\n",
       "      <td>Tier 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P09919</td>\n",
       "      <td>CSF3</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9NZ08</td>\n",
       "      <td>ERAP1</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P29459</td>\n",
       "      <td>IL12A</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q8TAD2</td>\n",
       "      <td>IL17D</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9UHD0</td>\n",
       "      <td>IL19</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q9HBE4</td>\n",
       "      <td>IL21</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>Q13007</td>\n",
       "      <td>IL24</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>P13232</td>\n",
       "      <td>IL7</td>\n",
       "      <td>Tier 3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>IL23R</td>\n",
       "      <td>Q5VWK5</td>\n",
       "      <td>O00421</td>\n",
       "      <td>CCRL2</td>\n",
       "      <td>Tier 3B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   g1.name p1.uniprot_id p2.uniprot_id  g2.name g2.druggability_tier\n",
       "0    IL23R        Q5VWK5        P04141     CSF2               Tier 1\n",
       "1    IL23R        Q5VWK5        P01562    IFNA1               Tier 1\n",
       "2    IL23R        Q5VWK5        P01579     IFNG               Tier 1\n",
       "3    IL23R        Q5VWK5        P22301     IL10               Tier 1\n",
       "4    IL23R        Q5VWK5        P29460    IL12B               Tier 1\n",
       "5    IL23R        Q5VWK5        P42701  IL12RB1               Tier 1\n",
       "6    IL23R        Q5VWK5        P35225     IL13               Tier 1\n",
       "7    IL23R        Q5VWK5        P40933     IL15               Tier 1\n",
       "8    IL23R        Q5VWK5        Q16552    IL17A               Tier 1\n",
       "9    IL23R        Q5VWK5        Q96PD4    IL17F               Tier 1\n",
       "10   IL23R        Q5VWK5        P60568      IL2               Tier 1\n",
       "11   IL23R        Q5VWK5        Q9GZX6     IL22               Tier 1\n",
       "12   IL23R        Q5VWK5        Q9NPF7    IL23A               Tier 1\n",
       "13   IL23R        Q5VWK5        P05112      IL4               Tier 1\n",
       "14   IL23R        Q5VWK5        P05113      IL5               Tier 1\n",
       "15   IL23R        Q5VWK5        P05231      IL6               Tier 1\n",
       "16   IL23R        Q5VWK5        P15248      IL9               Tier 1\n",
       "17   IL23R        Q5VWK5        P23458     JAK1               Tier 1\n",
       "18   IL23R        Q5VWK5        O60674     JAK2               Tier 1\n",
       "19   IL23R        Q5VWK5        P19838    NFKB1               Tier 1\n",
       "20   IL23R        Q5VWK5        P42336   PIK3CA               Tier 1\n",
       "21   IL23R        Q5VWK5        P51449     RORC               Tier 1\n",
       "22   IL23R        Q5VWK5        P40763    STAT3               Tier 1\n",
       "23   IL23R        Q5VWK5        Q969D9     TSLP               Tier 1\n",
       "24   IL23R        Q5VWK5        P29597     TYK2               Tier 1\n",
       "25   IL23R        Q5VWK5        P51684     CCR6               Tier 2\n",
       "26   IL23R        Q5VWK5        P25963   NFKBIA               Tier 2\n",
       "27   IL23R        Q5VWK5        Q9HC29     NOD2               Tier 2\n",
       "28   IL23R        Q5VWK5        P27986   PIK3R1               Tier 2\n",
       "29   IL23R        Q5VWK5        Q04206     RELA               Tier 2\n",
       "30   IL23R        Q5VWK5        P42224    STAT1               Tier 2\n",
       "31   IL23R        Q5VWK5        P42229   STAT5A               Tier 2\n",
       "32   IL23R        Q5VWK5        P42226    STAT6               Tier 2\n",
       "33   IL23R        Q5VWK5        P09919     CSF3              Tier 3A\n",
       "34   IL23R        Q5VWK5        Q9NZ08    ERAP1              Tier 3A\n",
       "35   IL23R        Q5VWK5        P29459    IL12A              Tier 3A\n",
       "36   IL23R        Q5VWK5        Q8TAD2    IL17D              Tier 3A\n",
       "37   IL23R        Q5VWK5        Q9UHD0     IL19              Tier 3A\n",
       "38   IL23R        Q5VWK5        Q9HBE4     IL21              Tier 3A\n",
       "39   IL23R        Q5VWK5        Q13007     IL24              Tier 3A\n",
       "40   IL23R        Q5VWK5        P13232      IL7              Tier 3A\n",
       "41   IL23R        Q5VWK5        O00421    CCRL2              Tier 3B"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_drug_targets_ppi(gene_name):\n",
    "    endpoint = \"/gene/druggability/ppi\"\n",
    "    url = f\"{API_URL}{endpoint}\"\n",
    "    params = {\"gene_name\": gene_name}\n",
    "    r = requests.get(url, params=params)\n",
    "    r.raise_for_status()\n",
    "    df = pd.json_normalize(r.json()[\"results\"])\n",
    "    return df\n",
    "\n",
    "\n",
    "GENE_NAME = \"IL23R\"\n",
    "OUTCOME_TRAIT = \"Inflammatory bowel disease\"\n",
    "ppi_df = get_drug_targets_ppi(gene_name=GENE_NAME)\n",
    "ppi_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfaae716-c5ed-4c81-a960-b69cfe5806f1",
   "metadata": {},
   "source": [
    "For further analysis we select the gene of interest (*IL23R*) as well as its interacting genes with Tier 1 druggability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4b496a27-08f3-47e9-8efe-5c42878e5c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IL23R',\n",
       " 'CSF2',\n",
       " 'IFNA1',\n",
       " 'IFNG',\n",
       " 'IL10',\n",
       " 'IL12B',\n",
       " 'IL12RB1',\n",
       " 'IL13',\n",
       " 'IL15',\n",
       " 'IL17A',\n",
       " 'IL17F',\n",
       " 'IL2',\n",
       " 'IL22',\n",
       " 'IL23A',\n",
       " 'IL4',\n",
       " 'IL5',\n",
       " 'IL6',\n",
       " 'IL9',\n",
       " 'JAK1',\n",
       " 'JAK2',\n",
       " 'NFKB1',\n",
       " 'PIK3CA',\n",
       " 'RORC',\n",
       " 'STAT3',\n",
       " 'TSLP',\n",
       " 'TYK2']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_gene_list(ppi_df, include_primary_gene: bool = True):\n",
    "    if include_primary_gene:\n",
    "        gene_list = list(ppi_df[\"g1.name\"].drop_duplicates()) + list(\n",
    "            ppi_df.query(\"`g2.druggability_tier` == 'Tier 1'\")[\"g2.name\"]\n",
    "        )\n",
    "    else:\n",
    "        gene_list = list(ppi_df.query(\"`g2.druggability_tier` == 'Tier 1'\")[\"g2.name\"])\n",
    "    return gene_list\n",
    "\n",
    "\n",
    "gene_list = get_gene_list(ppi_df)\n",
    "gene_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79e01e12-4c44-433f-b8b2-de9944e09f20",
   "metadata": {},
   "source": [
    "### 2. Looking for literature evidence\n",
    "EpiGraphDB facilitates fast processing of this information by allowing access to a host of literature-mined relationships that have been structured into semantic triples. These take the general form (subject, predicate, object) and have been generated using contemporary natural language processing techniques applied to a massive amount of published biomedical research papers by [SemMedDB](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR.html). In the following section we will query the API for the literature relationship between a given gene, *IL23R* (several studies confirmed IL23R associations in independent cohorts of patients with Crohn's disease or ulcerative colitis) and an outcome trait, Inflammatory bowel disease."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "78f9a98e-203e-4e44-a985-69baa8a06b0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>gene.name</th>\n",
       "      <th>lt.id</th>\n",
       "      <th>lt.name</th>\n",
       "      <th>lt.type</th>\n",
       "      <th>st.predicate</th>\n",
       "      <th>literature_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[23131344]</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[21155887, 17484863]</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>NEG_ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[31728561]</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[21155887, 18383521, 18383363, 25159710, 18341...</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[27852544]</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[21557945, 19030026]</td>\n",
       "      <td>CSF2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[17206685]</td>\n",
       "      <td>CSF2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[23891915]</td>\n",
       "      <td>IFNA1</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[24975266]</td>\n",
       "      <td>IFNA1</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[20951137, 28174758, 9836081]</td>\n",
       "      <td>IFNA1</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[19519446]</td>\n",
       "      <td>IFNG</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[3139380]</td>\n",
       "      <td>IFNG</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>[19740775, 18452147]</td>\n",
       "      <td>IFNG</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[10403730]</td>\n",
       "      <td>IFNG</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>[16573780, 27917223, 19184348, 28551707, 25999...</td>\n",
       "      <td>IL10</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>[27468578, 25296012]</td>\n",
       "      <td>IL10</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>[27468578]</td>\n",
       "      <td>IL10</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>[11271474]</td>\n",
       "      <td>IL10</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>NEG_PREDISPOSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>[24519095, 29023267, 17628614]</td>\n",
       "      <td>IL10</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[18383521, 23573954, 30541240, 22479607, 19817...</td>\n",
       "      <td>IL12B</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>[11023669, 22741617]</td>\n",
       "      <td>IL13</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>[9609761, 11023669]</td>\n",
       "      <td>IL15</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>[30193869, 21576383]</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>[30193869, 21994045, 18088064, 21576383]</td>\n",
       "      <td>IL17F</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>[6607860, 6237813, 1587419]</td>\n",
       "      <td>IL2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>[19201773]</td>\n",
       "      <td>IL22</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>[30193869, 27029486, 18753178, 18499066]</td>\n",
       "      <td>IL23A</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>[10477546]</td>\n",
       "      <td>IL4</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>[7806044, 8964392, 9389741]</td>\n",
       "      <td>IL4</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>[15766556]</td>\n",
       "      <td>IL6</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>[11204808]</td>\n",
       "      <td>IL6</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>NEG_ASSOCIATED_WITH</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>[25145003]</td>\n",
       "      <td>IL6</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>[11204808, 7683293]</td>\n",
       "      <td>IL6</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>[24120915]</td>\n",
       "      <td>IL6</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>[29788053]</td>\n",
       "      <td>IL9</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>[28652656, 11515847]</td>\n",
       "      <td>IL9</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>[31158699]</td>\n",
       "      <td>IL9</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>[31069840, 19817673, 20627814, 22269120]</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>[27852544]</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>[17600378, 9882195]</td>\n",
       "      <td>NFKB1</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>[21637825, 20004201]</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>[30006408]</td>\n",
       "      <td>RORC</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>[28770550]</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>[21733838]</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>[21631466, 25132422, 28785144, 19817673, 20627...</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>[27852544, 21994179]</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>[27697608]</td>\n",
       "      <td>TSLP</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>[21318591, 27697608]</td>\n",
       "      <td>TSLP</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>[26432894, 26432894]</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>C0021390</td>\n",
       "      <td>Inflammatory Bowel Diseases</td>\n",
       "      <td>[dsyn]</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            pubmed_id gene.name     lt.id  \\\n",
       "0                                          [23131344]     IL23R  C0021390   \n",
       "1                                [21155887, 17484863]     IL23R  C0021390   \n",
       "2                                          [31728561]     IL23R  C0021390   \n",
       "3   [21155887, 18383521, 18383363, 25159710, 18341...     IL23R  C0021390   \n",
       "4                                          [27852544]     IL23R  C0021390   \n",
       "5                                [21557945, 19030026]      CSF2  C0021390   \n",
       "6                                          [17206685]      CSF2  C0021390   \n",
       "7                                          [23891915]     IFNA1  C0021390   \n",
       "8                                          [24975266]     IFNA1  C0021390   \n",
       "9                       [20951137, 28174758, 9836081]     IFNA1  C0021390   \n",
       "10                                         [19519446]      IFNG  C0021390   \n",
       "11                                          [3139380]      IFNG  C0021390   \n",
       "12                               [19740775, 18452147]      IFNG  C0021390   \n",
       "13                                         [10403730]      IFNG  C0021390   \n",
       "14  [16573780, 27917223, 19184348, 28551707, 25999...      IL10  C0021390   \n",
       "15                               [27468578, 25296012]      IL10  C0021390   \n",
       "16                                         [27468578]      IL10  C0021390   \n",
       "17                                         [11271474]      IL10  C0021390   \n",
       "18                     [24519095, 29023267, 17628614]      IL10  C0021390   \n",
       "19  [18383521, 23573954, 30541240, 22479607, 19817...     IL12B  C0021390   \n",
       "20                               [11023669, 22741617]      IL13  C0021390   \n",
       "21                                [9609761, 11023669]      IL15  C0021390   \n",
       "22                               [30193869, 21576383]     IL17A  C0021390   \n",
       "23           [30193869, 21994045, 18088064, 21576383]     IL17F  C0021390   \n",
       "24                        [6607860, 6237813, 1587419]       IL2  C0021390   \n",
       "25                                         [19201773]      IL22  C0021390   \n",
       "26           [30193869, 27029486, 18753178, 18499066]     IL23A  C0021390   \n",
       "27                                         [10477546]       IL4  C0021390   \n",
       "28                        [7806044, 8964392, 9389741]       IL4  C0021390   \n",
       "29                                         [15766556]       IL6  C0021390   \n",
       "30                                         [11204808]       IL6  C0021390   \n",
       "31                                         [25145003]       IL6  C0021390   \n",
       "32                                [11204808, 7683293]       IL6  C0021390   \n",
       "33                                         [24120915]       IL6  C0021390   \n",
       "34                                         [29788053]       IL9  C0021390   \n",
       "35                               [28652656, 11515847]       IL9  C0021390   \n",
       "36                                         [31158699]       IL9  C0021390   \n",
       "37           [31069840, 19817673, 20627814, 22269120]      JAK2  C0021390   \n",
       "38                                         [27852544]      JAK2  C0021390   \n",
       "39                                [17600378, 9882195]     NFKB1  C0021390   \n",
       "40                               [21637825, 20004201]    PIK3CA  C0021390   \n",
       "41                                         [30006408]      RORC  C0021390   \n",
       "42                                         [28770550]     STAT3  C0021390   \n",
       "43                                         [21733838]     STAT3  C0021390   \n",
       "44  [21631466, 25132422, 28785144, 19817673, 20627...     STAT3  C0021390   \n",
       "45                               [27852544, 21994179]     STAT3  C0021390   \n",
       "46                                         [27697608]      TSLP  C0021390   \n",
       "47                               [21318591, 27697608]      TSLP  C0021390   \n",
       "48                               [26432894, 26432894]      TYK2  C0021390   \n",
       "\n",
       "                        lt.name lt.type         st.predicate  literature_count  \n",
       "0   Inflammatory Bowel Diseases  [dsyn]          PREDISPOSES                 1  \n",
       "1   Inflammatory Bowel Diseases  [dsyn]  NEG_ASSOCIATED_WITH                 2  \n",
       "2   Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "3   Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                21  \n",
       "4   Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "5   Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "6   Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "7   Inflammatory Bowel Diseases  [dsyn]               TREATS                 1  \n",
       "8   Inflammatory Bowel Diseases  [dsyn]             PREVENTS                 1  \n",
       "9   Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 3  \n",
       "10  Inflammatory Bowel Diseases  [dsyn]               TREATS                 1  \n",
       "11  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "12  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "13  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "14  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                13  \n",
       "15  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 2  \n",
       "16  Inflammatory Bowel Diseases  [dsyn]          PREDISPOSES                 1  \n",
       "17  Inflammatory Bowel Diseases  [dsyn]      NEG_PREDISPOSES                 1  \n",
       "18  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 3  \n",
       "19  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 6  \n",
       "20  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "21  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "22  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "23  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 4  \n",
       "24  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 3  \n",
       "25  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 1  \n",
       "26  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 4  \n",
       "27  Inflammatory Bowel Diseases  [dsyn]             DISRUPTS                 1  \n",
       "28  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 3  \n",
       "29  Inflammatory Bowel Diseases  [dsyn]               TREATS                 1  \n",
       "30  Inflammatory Bowel Diseases  [dsyn]  NEG_ASSOCIATED_WITH                 1  \n",
       "31  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "32  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "33  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "34  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "35  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "36  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "37  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 4  \n",
       "38  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 1  \n",
       "39  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "40  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "41  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 1  \n",
       "42  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "43  Inflammatory Bowel Diseases  [dsyn]             AUGMENTS                 1  \n",
       "44  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 9  \n",
       "45  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 2  \n",
       "46  Inflammatory Bowel Diseases  [dsyn]               CAUSES                 1  \n",
       "47  Inflammatory Bowel Diseases  [dsyn]      ASSOCIATED_WITH                 2  \n",
       "48  Inflammatory Bowel Diseases  [dsyn]              AFFECTS                 2  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_literature(outcome_trait, gene_list):\n",
    "    def per_gene(gene_name):\n",
    "        endpoint = \"/gene/literature\"\n",
    "        url = f\"{API_URL}{endpoint}\"\n",
    "        params = {\"gene_name\": gene_name, \"object_name\": outcome_trait.lower()}\n",
    "        r = requests.get(url, params=params)\n",
    "        try:\n",
    "            r.raise_for_status()\n",
    "            res_df = pd.json_normalize(r.json()[\"results\"])\n",
    "            if len(res_df) > 0:\n",
    "                res_df = res_df.assign(\n",
    "                    literature_count=lambda df: df[\"pubmed_id\"].apply(lambda x: len(x))\n",
    "                )\n",
    "            return res_df\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    res_df = pd.concat(\n",
    "        [per_gene(gene_name=gene_name) for gene_name in gene_list]\n",
    "    ).reset_index(drop=True)\n",
    "    return res_df\n",
    "\n",
    "\n",
    "literature_df = extract_literature(outcome_trait=OUTCOME_TRAIT, gene_list=gene_list)\n",
    "literature_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "346c905b-84bd-45fe-b619-2d8a989842db",
   "metadata": {},
   "source": [
    "### 3. Using Mendelian randomization results for causal effect estimation\n",
    "The next step is to find out whether any of these genes have a comparable and statistically plausable effect on IBD."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "45d581e8-e082-4024-b492-740b9bc8cc16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene.ensembl_id</th>\n",
       "      <th>gene.name</th>\n",
       "      <th>gwas.id</th>\n",
       "      <th>gwas.trait</th>\n",
       "      <th>r.beta</th>\n",
       "      <th>r.se</th>\n",
       "      <th>r.p</th>\n",
       "      <th>r.rsid</th>\n",
       "      <th>qtl_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000162594</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>1.500821</td>\n",
       "      <td>0.054592</td>\n",
       "      <td>2.212578e-166</td>\n",
       "      <td>rs11581607</td>\n",
       "      <td>pQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000113302</td>\n",
       "      <td>IL12B</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.417605</td>\n",
       "      <td>0.034490</td>\n",
       "      <td>9.590000e-34</td>\n",
       "      <td>rs4921484</td>\n",
       "      <td>pQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000162594</td>\n",
       "      <td>IL23R</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.886712</td>\n",
       "      <td>0.064420</td>\n",
       "      <td>4.165652e-43</td>\n",
       "      <td>rs2064689</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000164136</td>\n",
       "      <td>IL15</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>-1.421625</td>\n",
       "      <td>0.197131</td>\n",
       "      <td>5.530616e-13</td>\n",
       "      <td>rs75301646</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000113520</td>\n",
       "      <td>IL4</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.459848</td>\n",
       "      <td>0.084050</td>\n",
       "      <td>4.471537e-08</td>\n",
       "      <td>rs2070874</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>-1.896710</td>\n",
       "      <td>0.203808</td>\n",
       "      <td>1.322967e-20</td>\n",
       "      <td>rs4788084</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000109320</td>\n",
       "      <td>NFKB1</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.973556</td>\n",
       "      <td>0.173893</td>\n",
       "      <td>2.160849e-08</td>\n",
       "      <td>rs4766578</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000143365</td>\n",
       "      <td>RORC</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>-0.994991</td>\n",
       "      <td>0.116343</td>\n",
       "      <td>1.207271e-17</td>\n",
       "      <td>rs4845604</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000168610</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>ieu-a-294</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>0.597473</td>\n",
       "      <td>0.075700</td>\n",
       "      <td>2.958269e-15</td>\n",
       "      <td>rs1053004</td>\n",
       "      <td>eQTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene.ensembl_id gene.name    gwas.id                  gwas.trait    r.beta  \\\n",
       "0  ENSG00000162594     IL23R  ieu-a-294  Inflammatory bowel disease  1.500821   \n",
       "1  ENSG00000113302     IL12B  ieu-a-294  Inflammatory bowel disease  0.417605   \n",
       "2  ENSG00000162594     IL23R  ieu-a-294  Inflammatory bowel disease  0.886712   \n",
       "3  ENSG00000164136      IL15  ieu-a-294  Inflammatory bowel disease -1.421625   \n",
       "4  ENSG00000113520       IL4  ieu-a-294  Inflammatory bowel disease  0.459848   \n",
       "5  ENSG00000096968      JAK2  ieu-a-294  Inflammatory bowel disease -1.896710   \n",
       "6  ENSG00000109320     NFKB1  ieu-a-294  Inflammatory bowel disease  0.973556   \n",
       "7  ENSG00000143365      RORC  ieu-a-294  Inflammatory bowel disease -0.994991   \n",
       "8  ENSG00000168610     STAT3  ieu-a-294  Inflammatory bowel disease  0.597473   \n",
       "\n",
       "       r.se            r.p      r.rsid qtl_type  \n",
       "0  0.054592  2.212578e-166  rs11581607     pQTL  \n",
       "1  0.034490   9.590000e-34   rs4921484     pQTL  \n",
       "2  0.064420   4.165652e-43   rs2064689     eQTL  \n",
       "3  0.197131   5.530616e-13  rs75301646     eQTL  \n",
       "4  0.084050   4.471537e-08   rs2070874     eQTL  \n",
       "5  0.203808   1.322967e-20   rs4788084     eQTL  \n",
       "6  0.173893   2.160849e-08   rs4766578     eQTL  \n",
       "7  0.116343   1.207271e-17   rs4845604     eQTL  \n",
       "8  0.075700   2.958269e-15   rs1053004     eQTL  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_mr(outcome_trait, gene_list, qtl_type):\n",
    "    endpoint = \"/xqtl/single-snp-mr\"\n",
    "    url = f\"{API_URL}{endpoint}\"\n",
    "\n",
    "    def per_gene(gene_name):\n",
    "        params = {\n",
    "            \"exposure_gene\": gene_name,\n",
    "            \"outcome_trait\": outcome_trait,\n",
    "            \"qtl_type\": qtl_type,\n",
    "            \"pval_threshold\": 1e-5,\n",
    "        }\n",
    "        r = requests.get(url, params=params)\n",
    "        try:\n",
    "            r.raise_for_status()\n",
    "            df = pd.json_normalize(r.json()[\"results\"])\n",
    "            return df\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    res_df = pd.concat(\n",
    "        [per_gene(gene_name=gene_name) for gene_name in gene_list]\n",
    "    ).reset_index(drop=True)\n",
    "    return res_df\n",
    "\n",
    "\n",
    "# Search for both pqtl and eqtl\n",
    "xqtl_df = pd.concat(\n",
    "    [\n",
    "        extract_mr(\n",
    "            outcome_trait=OUTCOME_TRAIT, gene_list=gene_list, qtl_type=qtl_type\n",
    "        ).assign(qtl_type=qtl_type)\n",
    "        for qtl_type in [\"pQTL\", \"eQTL\"]\n",
    "    ]\n",
    ").reset_index(drop=True)\n",
    "xqtl_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1008f8c-a25f-483b-9dbe-91c365906106",
   "metadata": {},
   "source": [
    "### 4. Query metadata\n",
    "\n",
    "Here we query for the metadata information using the endpoint `GET /meta/schema`, which will be used for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b46775e-092b-4c65-ab8c-2ed5c4808276",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['nodes', 'edges', 'connections']) \n",
      "\n",
      "# nodes:\n",
      "{'Disease': {'count': 38960,\n",
      "             'properties': {'_id': {'indexed': True,\n",
      "                                    'type': 'STRING',\n",
      "                                    'unique': False},\n",
      "                            '_name': {'indexed': True,\n",
      "                                      'type': 'STRING',\n",
      "                                      'unique': False},\n",
      "                            '_source': {'indexed': False,\n",
      "                                        'type': 'LIST',\n",
      "                                        'unique': False},\n",
      "                            'definition': {'indexed': False,\n",
      "                                           'type': 'STRING',\n",
      "                                           'unique': False},\n",
      "                            'doid': {'indexed': True,\n",
      "                                     'type': 'LIST',\n",
      "                                     'unique': False},\n",
      "                            'efo': {'indexed': False,\n",
      "                                    'type': 'LIST',\n",
      "                     \n",
      "\n",
      "# edges:\n",
      "{'BIORXIV_OBJ': {'count': 32651,\n",
      "                 'properties': {'_source': {'array': True, 'type': 'LIST'}}},\n",
      " 'BIORXIV_PREDICATE': {'count': 32648,\n",
      "                       'properties': {'_source': {'array': True,\n",
      "                                                  'type': 'LIST'},\n",
      "                                      'count': {'array': False,\n",
      "                                                'type': 'INTEGER'},\n",
      "                                      'predicate': {'array': False,\n",
      "                                                    'type': 'STRING'}}},\n",
      " 'BIORXIV_SUB': {'count': 32657,\n",
      "                 'properties': {'_source': {'array': True, 'type': 'LIST'}}},\n",
      " 'BIORXIV_TO_LIT': {'count': 35211,\n",
      "                    'properties': {'_source': {'array': True, 'type': 'LIST'}}},\n",
      " 'CPIC': {'count': 375,\n",
      "          'properties': {'_source': {'array': True, 'type': 'LIST'},\n",
      "                         'cpic_level': {'array': False, 'type': 'STRING'},\n",
      "                         'guideline': {'array': F \n",
      "\n",
      "# connections:\n",
      "[{'count': 2461,\n",
      "  'from_node': 'Drug',\n",
      "  'rel': 'OPENTARGETS_DRUG_TO_DISEASE',\n",
      "  'to_node': 'Disease'},\n",
      " {'count': 5763,\n",
      "  'from_node': 'Gene',\n",
      "  'rel': 'GENE_TO_DISEASE',\n",
      "  'to_node': 'Disease'},\n",
      " {'count': 8247,\n",
      "  'from_node': 'Disease',\n",
      "  'rel': 'MONDO_MAP_UMLS',\n",
      "  'to_node': 'LiteratureTerm'},\n",
      " {'count': 2819,\n",
      "  'from_node': 'Disease',\n",
      "  'rel': 'MONDO_MAP_EFO',\n",
      "  'to_node': 'Efo'},\n",
      " {'count': 2463,\n",
      "  'from_node': 'Pathway',\n",
      "  'rel': 'PATHWAY_CHILD_OF',\n",
      "  'to_node': 'Pathway'},\n",
      " {'count': 121873,\n",
      "  'from_node': 'Protein',\n",
      "  'rel': 'PROTEIN_IN_PATHWAY',\n",
      "  'to_node': 'Pathway'},\n",
      " {'count': 1969,\n",
      "  'from_node': 'LiteratureTriple',\n",
      "  'rel': 'MEDRXIV_SUB',\n",
      "  'to_node': 'LiteratureTerm'},\n",
      " {'count': 5584547,\n",
      "  'from_node': 'LiteratureTerm',\n",
      "  'rel': 'SEMMEDDB_PREDICATE',\n",
      "  'to_node': 'LiteratureTerm'},\n",
      " {'count': 5584547,\n",
      "  'from_node': 'LiteratureTriple',\n",
      "  'rel': 'SEMMEDDB_SUB',\n",
      "  'to_node': 'LiteratureTerm'},\n",
      " {'count': 5556,\n",
      "  'from_node': 'Gwas',\n",
      "  'rel': 'METAMAP_LITE',\n",
      "  'to_node' \n",
      "\n"
     ]
    }
   ],
   "source": [
    "endpoint = \"/meta/schema\"\n",
    "params = {\"graphviz\": False, \"plot\": False}\n",
    "r = requests.get(f\"{API_URL}{endpoint}\", params=params)\n",
    "r.raise_for_status()\n",
    "metadata = r.json()\n",
    "\n",
    "# Preview of metadata information\n",
    "keys = metadata.keys()\n",
    "print(pformat(keys), \"\\n\")\n",
    "for key in list(keys):\n",
    "    print(f\"# {key}:\")\n",
    "    print(pformat(metadata[key])[:1000], \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddf956f3-f35f-4214-aaea-9f1495e860f3",
   "metadata": {},
   "source": [
    "We can extract the specific meta node information as a pandas dataframe from the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d488ef6-6ab8-4681-a20f-86165a62fec1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Disease</th>\n",
       "      <td>38,960</td>\n",
       "      <td>{'_name': {'type': 'STRING', 'indexed': True, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug</th>\n",
       "      <td>2,697</td>\n",
       "      <td>{'molecule_type': {'type': 'STRING', 'indexed'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Efo</th>\n",
       "      <td>25,390</td>\n",
       "      <td>{'_name': {'type': 'STRING', 'indexed': True, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <td>57,737</td>\n",
       "      <td>{'druggability_tier': {'type': 'STRING', 'inde...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gwas</th>\n",
       "      <td>34,494</td>\n",
       "      <td>{'note': {'type': 'STRING', 'indexed': False, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Literature</th>\n",
       "      <td>3,995,672</td>\n",
       "      <td>{'issn': {'type': 'STRING', 'indexed': False, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LiteratureTerm</th>\n",
       "      <td>108,905</td>\n",
       "      <td>{'_name': {'type': 'STRING', 'indexed': True, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LiteratureTriple</th>\n",
       "      <td>5,609,945</td>\n",
       "      <td>{'subject_id': {'type': 'STRING', 'indexed': T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathway</th>\n",
       "      <td>2,441</td>\n",
       "      <td>{'_name': {'type': 'STRING', 'indexed': True, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein</th>\n",
       "      <td>20,280</td>\n",
       "      <td>{'name': {'type': 'STRING', 'indexed': True, '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tissue</th>\n",
       "      <td>54</td>\n",
       "      <td>{'name': {'type': 'STRING', 'indexed': True, '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant</th>\n",
       "      <td>99,005</td>\n",
       "      <td>{'ref': {'type': 'STRING', 'indexed': False, '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      count                                         properties\n",
       "Disease              38,960  {'_name': {'type': 'STRING', 'indexed': True, ...\n",
       "Drug                  2,697  {'molecule_type': {'type': 'STRING', 'indexed'...\n",
       "Efo                  25,390  {'_name': {'type': 'STRING', 'indexed': True, ...\n",
       "Gene                 57,737  {'druggability_tier': {'type': 'STRING', 'inde...\n",
       "Gwas                 34,494  {'note': {'type': 'STRING', 'indexed': False, ...\n",
       "Literature        3,995,672  {'issn': {'type': 'STRING', 'indexed': False, ...\n",
       "LiteratureTerm      108,905  {'_name': {'type': 'STRING', 'indexed': True, ...\n",
       "LiteratureTriple  5,609,945  {'subject_id': {'type': 'STRING', 'indexed': T...\n",
       "Pathway               2,441  {'_name': {'type': 'STRING', 'indexed': True, ...\n",
       "Protein              20,280  {'name': {'type': 'STRING', 'indexed': True, '...\n",
       "Tissue                   54  {'name': {'type': 'STRING', 'indexed': True, '...\n",
       "Variant              99,005  {'ref': {'type': 'STRING', 'indexed': False, '..."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_node_df = pd.DataFrame.from_dict(metadata[\"nodes\"], orient=\"index\")\n",
    "\n",
    "(\n",
    "    meta_node_df.sort_index().assign(\n",
    "        count=lambda df: df[\"count\"].apply(lambda x: f\"{x:,}\")\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a8950f-4f85-4845-817c-f88452a75c04",
   "metadata": {},
   "source": [
    "We can also extract the meta relationship (edge) information, and the connections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd901d2c-6b1b-42b0-82f7-74105680fb7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>properties</th>\n",
       "      <th>from_node</th>\n",
       "      <th>to_node</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MONDO_MAP_EFO</th>\n",
       "      <td>2,819</td>\n",
       "      <td>{'_source': {'array': False, 'type': 'STRING'}}</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Efo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MONDO_MAP_UMLS</th>\n",
       "      <td>8,247</td>\n",
       "      <td>{'_source': {'array': False, 'type': 'STRING'}}</td>\n",
       "      <td>Disease</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OPENTARGETS_DRUG_TO_DISEASE</th>\n",
       "      <td>2,461</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPIC</th>\n",
       "      <td>375</td>\n",
       "      <td>{'pharmgkb_level_of_evidence': {'array': False...</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OPENTARGETS_DRUG_TO_TARGET</th>\n",
       "      <td>6,534</td>\n",
       "      <td>{'phase': {'array': False, 'type': 'STRING'}, ...</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EFO_CHILD_OF</th>\n",
       "      <td>43,132</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Efo</td>\n",
       "      <td>Efo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE_TO_DISEASE</th>\n",
       "      <td>5,763</td>\n",
       "      <td>{'last_updated': {'array': False, 'type': 'STR...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XQTL_MULTI_SNP_MR</th>\n",
       "      <td>3,015,233</td>\n",
       "      <td>{'p': {'array': False, 'type': 'FLOAT'}, 'se':...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XQTL_SINGLE_SNP_MR_GENE_GWAS</th>\n",
       "      <td>8,449,779</td>\n",
       "      <td>{'p': {'array': False, 'type': 'FLOAT'}, 'se':...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE_TO_PROTEIN</th>\n",
       "      <td>19,142</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EXPRESSED_IN</th>\n",
       "      <td>2,918,240</td>\n",
       "      <td>{'tpm': {'array': False, 'type': 'FLOAT'}, '_s...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_NLP_EFO</th>\n",
       "      <td>12,302</td>\n",
       "      <td>{'score': {'array': False, 'type': 'FLOAT'}, '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Efo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_EFO_EBI</th>\n",
       "      <td>281</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Efo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRS</th>\n",
       "      <td>118,124</td>\n",
       "      <td>{'p': {'array': False, 'type': 'FLOAT'}, 'r2':...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MR_EVE_MR</th>\n",
       "      <td>25,804,945</td>\n",
       "      <td>{'b': {'array': False, 'type': 'FLOAT'}, 'se':...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GEN_COR</th>\n",
       "      <td>840,960</td>\n",
       "      <td>{'h2_intercept_SE': {'array': False, 'type': '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OBS_COR</th>\n",
       "      <td>17,932</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}, '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_NLP</th>\n",
       "      <td>89,239,773</td>\n",
       "      <td>{'score': {'array': False, 'type': 'FLOAT'}, '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_TO_LITERATURE</th>\n",
       "      <td>28,111,669</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Literature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>METAMAP_LITE</th>\n",
       "      <td>5,556</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}, '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_TO_LITERATURE_TRIPLE</th>\n",
       "      <td>17,531,153</td>\n",
       "      <td>{'pval': {'array': False, 'type': 'FLOAT'}, 'g...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OPENGWAS_TOPHITS</th>\n",
       "      <td>160,283</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}, '...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GWAS_TO_VARIANT</th>\n",
       "      <td>26,436</td>\n",
       "      <td>{'se': {'array': False, 'type': 'FLOAT'}, 'nca...</td>\n",
       "      <td>Gwas</td>\n",
       "      <td>Variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TERM_TO_GENE</th>\n",
       "      <td>16,435</td>\n",
       "      <td>{'_source': {'array': False, 'type': 'STRING'}}</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEMMEDDB_PREDICATE</th>\n",
       "      <td>5,584,547</td>\n",
       "      <td>{'count': {'array': False, 'type': 'INTEGER'},...</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIORXIV_PREDICATE</th>\n",
       "      <td>32,648</td>\n",
       "      <td>{'count': {'array': False, 'type': 'INTEGER'},...</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEDRXIV_PREDICATE</th>\n",
       "      <td>1,969</td>\n",
       "      <td>{'count': {'array': False, 'type': 'INTEGER'},...</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIORXIV_TO_LIT</th>\n",
       "      <td>35,211</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>Literature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEMMEDDB_TO_LIT</th>\n",
       "      <td>10,589,785</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>Literature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEDRXIV_TO_LIT</th>\n",
       "      <td>2,091</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>Literature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEDRXIV_SUB</th>\n",
       "      <td>1,969</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEMMEDDB_SUB</th>\n",
       "      <td>5,584,547</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEDRXIV_OBJ</th>\n",
       "      <td>1,969</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIORXIV_OBJ</th>\n",
       "      <td>32,651</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIORXIV_SUB</th>\n",
       "      <td>32,657</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEMMEDDB_OBJ</th>\n",
       "      <td>5,584,547</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>LiteratureTriple</td>\n",
       "      <td>LiteratureTerm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PATHWAY_CHILD_OF</th>\n",
       "      <td>2,463</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>Pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROTEIN_IN_PATHWAY</th>\n",
       "      <td>121,873</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Protein</td>\n",
       "      <td>Pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRING_INTERACT_WITH</th>\n",
       "      <td>827,184</td>\n",
       "      <td>{'score': {'array': False, 'type': 'FLOAT'}, '...</td>\n",
       "      <td>Protein</td>\n",
       "      <td>Protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VARIANT_TO_GENE</th>\n",
       "      <td>108,561</td>\n",
       "      <td>{'amino_acids': {'array': False, 'type': 'STRI...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XQTL_SINGLE_SNP_MR_SNP_GENE</th>\n",
       "      <td>41,564</td>\n",
       "      <td>{'_source': {'array': True, 'type': 'LIST'}}</td>\n",
       "      <td>Variant</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   count  \\\n",
       "MONDO_MAP_EFO                      2,819   \n",
       "MONDO_MAP_UMLS                     8,247   \n",
       "OPENTARGETS_DRUG_TO_DISEASE        2,461   \n",
       "CPIC                                 375   \n",
       "OPENTARGETS_DRUG_TO_TARGET         6,534   \n",
       "EFO_CHILD_OF                      43,132   \n",
       "GENE_TO_DISEASE                    5,763   \n",
       "XQTL_MULTI_SNP_MR              3,015,233   \n",
       "XQTL_SINGLE_SNP_MR_GENE_GWAS   8,449,779   \n",
       "GENE_TO_PROTEIN                   19,142   \n",
       "EXPRESSED_IN                   2,918,240   \n",
       "GWAS_NLP_EFO                      12,302   \n",
       "GWAS_EFO_EBI                         281   \n",
       "PRS                              118,124   \n",
       "MR_EVE_MR                     25,804,945   \n",
       "GEN_COR                          840,960   \n",
       "OBS_COR                           17,932   \n",
       "GWAS_NLP                      89,239,773   \n",
       "GWAS_TO_LITERATURE            28,111,669   \n",
       "METAMAP_LITE                       5,556   \n",
       "GWAS_TO_LITERATURE_TRIPLE     17,531,153   \n",
       "OPENGWAS_TOPHITS                 160,283   \n",
       "GWAS_TO_VARIANT                   26,436   \n",
       "TERM_TO_GENE                      16,435   \n",
       "SEMMEDDB_PREDICATE             5,584,547   \n",
       "BIORXIV_PREDICATE                 32,648   \n",
       "MEDRXIV_PREDICATE                  1,969   \n",
       "BIORXIV_TO_LIT                    35,211   \n",
       "SEMMEDDB_TO_LIT               10,589,785   \n",
       "MEDRXIV_TO_LIT                     2,091   \n",
       "MEDRXIV_SUB                        1,969   \n",
       "SEMMEDDB_SUB                   5,584,547   \n",
       "MEDRXIV_OBJ                        1,969   \n",
       "BIORXIV_OBJ                       32,651   \n",
       "BIORXIV_SUB                       32,657   \n",
       "SEMMEDDB_OBJ                   5,584,547   \n",
       "PATHWAY_CHILD_OF                   2,463   \n",
       "PROTEIN_IN_PATHWAY               121,873   \n",
       "STRING_INTERACT_WITH             827,184   \n",
       "VARIANT_TO_GENE                  108,561   \n",
       "XQTL_SINGLE_SNP_MR_SNP_GENE       41,564   \n",
       "\n",
       "                                                                     properties  \\\n",
       "MONDO_MAP_EFO                   {'_source': {'array': False, 'type': 'STRING'}}   \n",
       "MONDO_MAP_UMLS                  {'_source': {'array': False, 'type': 'STRING'}}   \n",
       "OPENTARGETS_DRUG_TO_DISEASE        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "CPIC                          {'pharmgkb_level_of_evidence': {'array': False...   \n",
       "OPENTARGETS_DRUG_TO_TARGET    {'phase': {'array': False, 'type': 'STRING'}, ...   \n",
       "EFO_CHILD_OF                       {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "GENE_TO_DISEASE               {'last_updated': {'array': False, 'type': 'STR...   \n",
       "XQTL_MULTI_SNP_MR             {'p': {'array': False, 'type': 'FLOAT'}, 'se':...   \n",
       "XQTL_SINGLE_SNP_MR_GENE_GWAS  {'p': {'array': False, 'type': 'FLOAT'}, 'se':...   \n",
       "GENE_TO_PROTEIN                    {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "EXPRESSED_IN                  {'tpm': {'array': False, 'type': 'FLOAT'}, '_s...   \n",
       "GWAS_NLP_EFO                  {'score': {'array': False, 'type': 'FLOAT'}, '...   \n",
       "GWAS_EFO_EBI                       {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "PRS                           {'p': {'array': False, 'type': 'FLOAT'}, 'r2':...   \n",
       "MR_EVE_MR                     {'b': {'array': False, 'type': 'FLOAT'}, 'se':...   \n",
       "GEN_COR                       {'h2_intercept_SE': {'array': False, 'type': '...   \n",
       "OBS_COR                       {'_source': {'array': True, 'type': 'LIST'}, '...   \n",
       "GWAS_NLP                      {'score': {'array': False, 'type': 'FLOAT'}, '...   \n",
       "GWAS_TO_LITERATURE                 {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "METAMAP_LITE                  {'_source': {'array': True, 'type': 'LIST'}, '...   \n",
       "GWAS_TO_LITERATURE_TRIPLE     {'pval': {'array': False, 'type': 'FLOAT'}, 'g...   \n",
       "OPENGWAS_TOPHITS              {'_source': {'array': True, 'type': 'LIST'}, '...   \n",
       "GWAS_TO_VARIANT               {'se': {'array': False, 'type': 'FLOAT'}, 'nca...   \n",
       "TERM_TO_GENE                    {'_source': {'array': False, 'type': 'STRING'}}   \n",
       "SEMMEDDB_PREDICATE            {'count': {'array': False, 'type': 'INTEGER'},...   \n",
       "BIORXIV_PREDICATE             {'count': {'array': False, 'type': 'INTEGER'},...   \n",
       "MEDRXIV_PREDICATE             {'count': {'array': False, 'type': 'INTEGER'},...   \n",
       "BIORXIV_TO_LIT                     {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "SEMMEDDB_TO_LIT                    {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "MEDRXIV_TO_LIT                     {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "MEDRXIV_SUB                        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "SEMMEDDB_SUB                       {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "MEDRXIV_OBJ                        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "BIORXIV_OBJ                        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "BIORXIV_SUB                        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "SEMMEDDB_OBJ                       {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "PATHWAY_CHILD_OF                   {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "PROTEIN_IN_PATHWAY                 {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "STRING_INTERACT_WITH          {'score': {'array': False, 'type': 'FLOAT'}, '...   \n",
       "VARIANT_TO_GENE               {'amino_acids': {'array': False, 'type': 'STRI...   \n",
       "XQTL_SINGLE_SNP_MR_SNP_GENE        {'_source': {'array': True, 'type': 'LIST'}}   \n",
       "\n",
       "                                     from_node           to_node  \n",
       "MONDO_MAP_EFO                          Disease               Efo  \n",
       "MONDO_MAP_UMLS                         Disease    LiteratureTerm  \n",
       "OPENTARGETS_DRUG_TO_DISEASE               Drug           Disease  \n",
       "CPIC                                      Drug              Gene  \n",
       "OPENTARGETS_DRUG_TO_TARGET                Drug              Gene  \n",
       "EFO_CHILD_OF                               Efo               Efo  \n",
       "GENE_TO_DISEASE                           Gene           Disease  \n",
       "XQTL_MULTI_SNP_MR                         Gene              Gwas  \n",
       "XQTL_SINGLE_SNP_MR_GENE_GWAS              Gene              Gwas  \n",
       "GENE_TO_PROTEIN                           Gene           Protein  \n",
       "EXPRESSED_IN                              Gene            Tissue  \n",
       "GWAS_NLP_EFO                              Gwas               Efo  \n",
       "GWAS_EFO_EBI                              Gwas               Efo  \n",
       "PRS                                       Gwas              Gwas  \n",
       "MR_EVE_MR                                 Gwas              Gwas  \n",
       "GEN_COR                                   Gwas              Gwas  \n",
       "OBS_COR                                   Gwas              Gwas  \n",
       "GWAS_NLP                                  Gwas              Gwas  \n",
       "GWAS_TO_LITERATURE                        Gwas        Literature  \n",
       "METAMAP_LITE                              Gwas    LiteratureTerm  \n",
       "GWAS_TO_LITERATURE_TRIPLE                 Gwas  LiteratureTriple  \n",
       "OPENGWAS_TOPHITS                          Gwas           Variant  \n",
       "GWAS_TO_VARIANT                           Gwas           Variant  \n",
       "TERM_TO_GENE                    LiteratureTerm              Gene  \n",
       "SEMMEDDB_PREDICATE              LiteratureTerm    LiteratureTerm  \n",
       "BIORXIV_PREDICATE               LiteratureTerm    LiteratureTerm  \n",
       "MEDRXIV_PREDICATE               LiteratureTerm    LiteratureTerm  \n",
       "BIORXIV_TO_LIT                LiteratureTriple        Literature  \n",
       "SEMMEDDB_TO_LIT               LiteratureTriple        Literature  \n",
       "MEDRXIV_TO_LIT                LiteratureTriple        Literature  \n",
       "MEDRXIV_SUB                   LiteratureTriple    LiteratureTerm  \n",
       "SEMMEDDB_SUB                  LiteratureTriple    LiteratureTerm  \n",
       "MEDRXIV_OBJ                   LiteratureTriple    LiteratureTerm  \n",
       "BIORXIV_OBJ                   LiteratureTriple    LiteratureTerm  \n",
       "BIORXIV_SUB                   LiteratureTriple    LiteratureTerm  \n",
       "SEMMEDDB_OBJ                  LiteratureTriple    LiteratureTerm  \n",
       "PATHWAY_CHILD_OF                       Pathway           Pathway  \n",
       "PROTEIN_IN_PATHWAY                     Protein           Pathway  \n",
       "STRING_INTERACT_WITH                   Protein           Protein  \n",
       "VARIANT_TO_GENE                        Variant              Gene  \n",
       "XQTL_SINGLE_SNP_MR_SNP_GENE            Variant              Gene  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_rel_df = pd.DataFrame.from_dict(metadata[\"edges\"], orient=\"index\").merge(\n",
    "    pd.DataFrame.from_dict(\n",
    "        {_[\"rel\"]: _ for _ in metadata[\"connections\"]}, orient=\"index\"\n",
    "    )[[\"from_node\", \"to_node\"]],\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    ")\n",
    "\n",
    "(\n",
    "    meta_rel_df.sort_values(by=[\"from_node\", \"to_node\"]).assign(\n",
    "        count=lambda df: df[\"count\"].apply(lambda x: f\"{x:,}\")\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc357862-1934-4ebe-8678-de96afe93a2e",
   "metadata": {},
   "source": [
    "We can generate a network diagram of the graph db schema using `networkx`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5413d114-1140-4445-b832-dd7142c22917",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg height=\"558pt\" version=\"1.1\" viewBox=\"0 0 572.4 558\" width=\"572.4pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-08-05T04:37:23.066938</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.4.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linecap:butt;stroke-linejoin:round;}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 558 \n",
       "L 572.4 558 \n",
       "L 572.4 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill:#ffffff;\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"LineCollection_1\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 325.923451 428.254354 \n",
       "L 235.951357 408.420739 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 325.923451 428.254354 \n",
       "L 292.5338 311.548002 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 235.951357 408.420739 \n",
       "L 292.5338 311.548002 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 235.951357 408.420739 \n",
       "L 214.030034 215.97233 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 235.951357 408.420739 \n",
       "L 128.600619 434.469284 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 127.768172 109.411919 \n",
       "L 214.030034 215.97233 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 127.768172 109.411919 \n",
       "L 150.365414 251.982801 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 127.768172 109.411919 \n",
       "L 55.621488 134.472991 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 214.030034 215.97233 \n",
       "L 214.030034 215.97233 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 214.030034 215.97233 \n",
       "L 150.365414 251.982801 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 214.030034 215.97233 \n",
       "L 292.5338 311.548002 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 292.5338 311.548002 \n",
       "L 196.257317 339.243017 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 292.5338 311.548002 \n",
       "L 370.671866 504.604743 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 292.5338 311.548002 \n",
       "L 150.365414 251.982801 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 292.5338 311.548002 \n",
       "L 422.049451 176.361536 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 150.365414 251.982801 \n",
       "L 128.600619 434.469284 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 150.365414 251.982801 \n",
       "L 150.365414 251.982801 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 150.365414 251.982801 \n",
       "L 196.257317 339.243017 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 150.365414 251.982801 \n",
       "L 55.621488 134.472991 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 128.600619 434.469284 \n",
       "L 128.600619 434.469284 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 422.049451 176.361536 \n",
       "L 422.049451 176.361536 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 422.049451 176.361536 \n",
       "L 516.778512 74.881423 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 516.778512 74.881423 \n",
       "L 516.778512 74.881423 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_2\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 214.030034 172.999998 \n",
       "C 226.456399 172.999998 226.456399 215.97233 214.030034 215.97233 \n",
       "C 201.603669 215.97233 201.603669 172.999998 214.030034 172.999998 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-linecap:round;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 150.365414 209.010469 \n",
       "C 162.791779 209.010469 162.791779 251.982801 150.365414 251.982801 \n",
       "C 137.939048 251.982801 137.939048 209.010469 150.365414 209.010469 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-linecap:round;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 128.600619 391.496952 \n",
       "C 141.026984 391.496952 141.026984 434.469284 128.600619 434.469284 \n",
       "C 116.174254 434.469284 116.174254 391.496952 128.600619 391.496952 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-linecap:round;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 422.049451 133.389204 \n",
       "C 434.475816 133.389204 434.475816 176.361536 422.049451 176.361536 \n",
       "C 409.623085 176.361536 409.623085 133.389204 422.049451 133.389204 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-linecap:round;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path clip-path=\"url(#pa70d908b59)\" d=\"M 516.778512 31.909091 \n",
       "C 529.204878 31.909091 529.204878 74.881423 516.778512 74.881423 \n",
       "C 504.352147 74.881423 504.352147 31.909091 516.778512 31.909091 \n",
       "\" style=\"fill:none;stroke:#000000;stroke-linecap:round;stroke-width:0.75;\"/>\n",
       "   </g>\n",
       "   <g id=\"PathCollection_1\">\n",
       "    <defs>\n",
       "     <path d=\"M 0 27.386128 \n",
       "C 7.262886 27.386128 14.229281 24.500553 19.364917 19.364917 \n",
       "C 24.500553 14.229281 27.386128 7.262886 27.386128 0 \n",
       "C 27.386128 -7.262886 24.500553 -14.229281 19.364917 -19.364917 \n",
       "C 14.229281 -24.500553 7.262886 -27.386128 0 -27.386128 \n",
       "C -7.262886 -27.386128 -14.229281 -24.500553 -19.364917 -19.364917 \n",
       "C -24.500553 -14.229281 -27.386128 -7.262886 -27.386128 0 \n",
       "C -27.386128 7.262886 -24.500553 14.229281 -19.364917 19.364917 \n",
       "C -14.229281 24.500553 -7.262886 27.386128 0 27.386128 \n",
       "z\n",
       "\" id=\"m5192ea9bf9\" style=\"stroke:#808080;\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"325.923451\" xlink:href=\"#m5192ea9bf9\" y=\"428.254354\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"235.951357\" xlink:href=\"#m5192ea9bf9\" y=\"408.420739\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"127.768172\" xlink:href=\"#m5192ea9bf9\" y=\"109.411919\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"214.030034\" xlink:href=\"#m5192ea9bf9\" y=\"215.97233\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"292.5338\" xlink:href=\"#m5192ea9bf9\" y=\"311.548002\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"150.365414\" xlink:href=\"#m5192ea9bf9\" y=\"251.982801\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"128.600619\" xlink:href=\"#m5192ea9bf9\" y=\"434.469284\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"422.049451\" xlink:href=\"#m5192ea9bf9\" y=\"176.361536\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"196.257317\" xlink:href=\"#m5192ea9bf9\" y=\"339.243017\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"370.671866\" xlink:href=\"#m5192ea9bf9\" y=\"504.604743\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"55.621488\" xlink:href=\"#m5192ea9bf9\" y=\"134.472991\"/>\n",
       "     <use style=\"fill:#87ceeb;stroke:#808080;\" x=\"516.778512\" xlink:href=\"#m5192ea9bf9\" y=\"74.881423\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Drug -->\n",
       "     <g transform=\"translate(312.168764 430.935604)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 1791 3756 \n",
       "L 1791 909 \n",
       "L 2222 909 \n",
       "Q 2959 909 3348 1275 \n",
       "Q 3738 1641 3738 2338 \n",
       "Q 3738 3031 3350 3393 \n",
       "Q 2963 3756 2222 3756 \n",
       "L 1791 3756 \n",
       "z\n",
       "M 588 4666 \n",
       "L 1856 4666 \n",
       "Q 2919 4666 3439 4514 \n",
       "Q 3959 4363 4331 4000 \n",
       "Q 4659 3684 4818 3271 \n",
       "Q 4978 2859 4978 2338 \n",
       "Q 4978 1809 4818 1395 \n",
       "Q 4659 981 4331 666 \n",
       "Q 3956 303 3431 151 \n",
       "Q 2906 0 1856 0 \n",
       "L 588 0 \n",
       "L 588 4666 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-44\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 3138 2547 \n",
       "Q 2991 2616 2845 2648 \n",
       "Q 2700 2681 2553 2681 \n",
       "Q 2122 2681 1889 2404 \n",
       "Q 1656 2128 1656 1613 \n",
       "L 1656 0 \n",
       "L 538 0 \n",
       "L 538 3500 \n",
       "L 1656 3500 \n",
       "L 1656 2925 \n",
       "Q 1872 3269 2151 3426 \n",
       "Q 2431 3584 2822 3584 \n",
       "Q 2878 3584 2943 3579 \n",
       "Q 3009 3575 3134 3559 \n",
       "L 3138 2547 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-72\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 500 1363 \n",
       "L 500 3500 \n",
       "L 1625 3500 \n",
       "L 1625 3150 \n",
       "Q 1625 2866 1622 2436 \n",
       "Q 1619 2006 1619 1863 \n",
       "Q 1619 1441 1641 1255 \n",
       "Q 1663 1069 1716 984 \n",
       "Q 1784 875 1895 815 \n",
       "Q 2006 756 2150 756 \n",
       "Q 2500 756 2700 1025 \n",
       "Q 2900 1294 2900 1772 \n",
       "L 2900 3500 \n",
       "L 4019 3500 \n",
       "L 4019 0 \n",
       "L 2900 0 \n",
       "L 2900 506 \n",
       "Q 2647 200 2364 54 \n",
       "Q 2081 -91 1741 -91 \n",
       "Q 1134 -91 817 281 \n",
       "Q 500 653 500 1363 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-75\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 2919 594 \n",
       "Q 2688 288 2409 144 \n",
       "Q 2131 0 1766 0 \n",
       "Q 1125 0 706 504 \n",
       "Q 288 1009 288 1791 \n",
       "Q 288 2575 706 3076 \n",
       "Q 1125 3578 1766 3578 \n",
       "Q 2131 3578 2409 3434 \n",
       "Q 2688 3291 2919 2981 \n",
       "L 2919 3500 \n",
       "L 4044 3500 \n",
       "L 4044 353 \n",
       "Q 4044 -491 3511 -936 \n",
       "Q 2978 -1381 1966 -1381 \n",
       "Q 1638 -1381 1331 -1331 \n",
       "Q 1025 -1281 716 -1178 \n",
       "L 716 -306 \n",
       "Q 1009 -475 1290 -558 \n",
       "Q 1572 -641 1856 -641 \n",
       "Q 2406 -641 2662 -400 \n",
       "Q 2919 -159 2919 353 \n",
       "L 2919 594 \n",
       "z\n",
       "M 2181 2772 \n",
       "Q 1834 2772 1640 2515 \n",
       "Q 1447 2259 1447 1791 \n",
       "Q 1447 1309 1634 1061 \n",
       "Q 1822 813 2181 813 \n",
       "Q 2531 813 2725 1069 \n",
       "Q 2919 1325 2919 1791 \n",
       "Q 2919 2259 2725 2515 \n",
       "Q 2531 2772 2181 2772 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-67\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-44\"/>\n",
       "      <use x=\"83.007812\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"132.324219\" xlink:href=\"#DejaVuSans-Bold-75\"/>\n",
       "      <use x=\"203.515625\" xlink:href=\"#DejaVuSans-Bold-67\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Disease -->\n",
       "     <g transform=\"translate(213.977919 411.180114)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 538 3500 \n",
       "L 1656 3500 \n",
       "L 1656 0 \n",
       "L 538 0 \n",
       "L 538 3500 \n",
       "z\n",
       "M 538 4863 \n",
       "L 1656 4863 \n",
       "L 1656 3950 \n",
       "L 538 3950 \n",
       "L 538 4863 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-69\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 3272 3391 \n",
       "L 3272 2541 \n",
       "Q 2913 2691 2578 2766 \n",
       "Q 2244 2841 1947 2841 \n",
       "Q 1628 2841 1473 2761 \n",
       "Q 1319 2681 1319 2516 \n",
       "Q 1319 2381 1436 2309 \n",
       "Q 1553 2238 1856 2203 \n",
       "L 2053 2175 \n",
       "Q 2913 2066 3209 1816 \n",
       "Q 3506 1566 3506 1031 \n",
       "Q 3506 472 3093 190 \n",
       "Q 2681 -91 1863 -91 \n",
       "Q 1516 -91 1145 -36 \n",
       "Q 775 19 384 128 \n",
       "L 384 978 \n",
       "Q 719 816 1070 734 \n",
       "Q 1422 653 1784 653 \n",
       "Q 2113 653 2278 743 \n",
       "Q 2444 834 2444 1013 \n",
       "Q 2444 1163 2330 1236 \n",
       "Q 2216 1309 1875 1350 \n",
       "L 1678 1375 \n",
       "Q 931 1469 631 1722 \n",
       "Q 331 1975 331 2491 \n",
       "Q 331 3047 712 3315 \n",
       "Q 1094 3584 1881 3584 \n",
       "Q 2191 3584 2531 3537 \n",
       "Q 2872 3491 3272 3391 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-73\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 4031 1759 \n",
       "L 4031 1441 \n",
       "L 1416 1441 \n",
       "Q 1456 1047 1700 850 \n",
       "Q 1944 653 2381 653 \n",
       "Q 2734 653 3104 758 \n",
       "Q 3475 863 3866 1075 \n",
       "L 3866 213 \n",
       "Q 3469 63 3072 -14 \n",
       "Q 2675 -91 2278 -91 \n",
       "Q 1328 -91 801 392 \n",
       "Q 275 875 275 1747 \n",
       "Q 275 2603 792 3093 \n",
       "Q 1309 3584 2216 3584 \n",
       "Q 3041 3584 3536 3087 \n",
       "Q 4031 2591 4031 1759 \n",
       "z\n",
       "M 2881 2131 \n",
       "Q 2881 2450 2695 2645 \n",
       "Q 2509 2841 2209 2841 \n",
       "Q 1884 2841 1681 2658 \n",
       "Q 1478 2475 1428 2131 \n",
       "L 2881 2131 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-65\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 2106 1575 \n",
       "Q 1756 1575 1579 1456 \n",
       "Q 1403 1338 1403 1106 \n",
       "Q 1403 894 1545 773 \n",
       "Q 1688 653 1941 653 \n",
       "Q 2256 653 2472 879 \n",
       "Q 2688 1106 2688 1447 \n",
       "L 2688 1575 \n",
       "L 2106 1575 \n",
       "z\n",
       "M 3816 1997 \n",
       "L 3816 0 \n",
       "L 2688 0 \n",
       "L 2688 519 \n",
       "Q 2463 200 2181 54 \n",
       "Q 1900 -91 1497 -91 \n",
       "Q 953 -91 614 226 \n",
       "Q 275 544 275 1050 \n",
       "Q 275 1666 698 1953 \n",
       "Q 1122 2241 2028 2241 \n",
       "L 2688 2241 \n",
       "L 2688 2328 \n",
       "Q 2688 2594 2478 2717 \n",
       "Q 2269 2841 1825 2841 \n",
       "Q 1466 2841 1156 2769 \n",
       "Q 847 2697 581 2553 \n",
       "L 581 3406 \n",
       "Q 941 3494 1303 3539 \n",
       "Q 1666 3584 2028 3584 \n",
       "Q 2975 3584 3395 3211 \n",
       "Q 3816 2838 3816 1997 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-61\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-44\"/>\n",
       "      <use x=\"83.007812\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"117.285156\" xlink:href=\"#DejaVuSans-Bold-73\"/>\n",
       "      <use x=\"176.806641\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"244.628906\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"312.109375\" xlink:href=\"#DejaVuSans-Bold-73\"/>\n",
       "      <use x=\"371.630859\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- LiteratureTriple -->\n",
       "     <g transform=\"translate(83.715828 112.171294)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 588 4666 \n",
       "L 1791 4666 \n",
       "L 1791 909 \n",
       "L 3903 909 \n",
       "L 3903 0 \n",
       "L 588 0 \n",
       "L 588 4666 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-4c\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 1759 4494 \n",
       "L 1759 3500 \n",
       "L 2913 3500 \n",
       "L 2913 2700 \n",
       "L 1759 2700 \n",
       "L 1759 1216 \n",
       "Q 1759 972 1856 886 \n",
       "Q 1953 800 2241 800 \n",
       "L 2816 800 \n",
       "L 2816 0 \n",
       "L 1856 0 \n",
       "Q 1194 0 917 276 \n",
       "Q 641 553 641 1216 \n",
       "L 641 2700 \n",
       "L 84 2700 \n",
       "L 84 3500 \n",
       "L 641 3500 \n",
       "L 641 4494 \n",
       "L 1759 4494 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-74\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 31 4666 \n",
       "L 4331 4666 \n",
       "L 4331 3756 \n",
       "L 2784 3756 \n",
       "L 2784 0 \n",
       "L 1581 0 \n",
       "L 1581 3756 \n",
       "L 31 3756 \n",
       "L 31 4666 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-54\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 1656 506 \n",
       "L 1656 -1331 \n",
       "L 538 -1331 \n",
       "L 538 3500 \n",
       "L 1656 3500 \n",
       "L 1656 2988 \n",
       "Q 1888 3294 2169 3439 \n",
       "Q 2450 3584 2816 3584 \n",
       "Q 3463 3584 3878 3070 \n",
       "Q 4294 2556 4294 1747 \n",
       "Q 4294 938 3878 423 \n",
       "Q 3463 -91 2816 -91 \n",
       "Q 2450 -91 2169 54 \n",
       "Q 1888 200 1656 506 \n",
       "z\n",
       "M 2400 2772 \n",
       "Q 2041 2772 1848 2508 \n",
       "Q 1656 2244 1656 1747 \n",
       "Q 1656 1250 1848 986 \n",
       "Q 2041 722 2400 722 \n",
       "Q 2759 722 2948 984 \n",
       "Q 3138 1247 3138 1747 \n",
       "Q 3138 2247 2948 2509 \n",
       "Q 2759 2772 2400 2772 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-70\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 538 4863 \n",
       "L 1656 4863 \n",
       "L 1656 0 \n",
       "L 538 0 \n",
       "L 538 4863 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-6c\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-4c\"/>\n",
       "      <use x=\"63.720703\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"97.998047\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"145.800781\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"213.623047\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"262.939453\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"330.419922\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"378.222656\" xlink:href=\"#DejaVuSans-Bold-75\"/>\n",
       "      <use x=\"449.414062\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"498.730469\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"566.552734\" xlink:href=\"#DejaVuSans-Bold-54\"/>\n",
       "      <use x=\"623.765625\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"673.082031\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"707.359375\" xlink:href=\"#DejaVuSans-Bold-70\"/>\n",
       "      <use x=\"778.941406\" xlink:href=\"#DejaVuSans-Bold-6c\"/>\n",
       "      <use x=\"813.21875\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- LiteratureTerm -->\n",
       "     <g transform=\"translate(171.887065 218.731705)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 3781 2919 \n",
       "Q 3994 3244 4286 3414 \n",
       "Q 4578 3584 4928 3584 \n",
       "Q 5531 3584 5847 3212 \n",
       "Q 6163 2841 6163 2131 \n",
       "L 6163 0 \n",
       "L 5038 0 \n",
       "L 5038 1825 \n",
       "Q 5041 1866 5042 1909 \n",
       "Q 5044 1953 5044 2034 \n",
       "Q 5044 2406 4934 2573 \n",
       "Q 4825 2741 4581 2741 \n",
       "Q 4263 2741 4089 2478 \n",
       "Q 3916 2216 3909 1719 \n",
       "L 3909 0 \n",
       "L 2784 0 \n",
       "L 2784 1825 \n",
       "Q 2784 2406 2684 2573 \n",
       "Q 2584 2741 2328 2741 \n",
       "Q 2006 2741 1831 2477 \n",
       "Q 1656 2213 1656 1722 \n",
       "L 1656 0 \n",
       "L 531 0 \n",
       "L 531 3500 \n",
       "L 1656 3500 \n",
       "L 1656 2988 \n",
       "Q 1863 3284 2130 3434 \n",
       "Q 2397 3584 2719 3584 \n",
       "Q 3081 3584 3359 3409 \n",
       "Q 3638 3234 3781 2919 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-6d\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-4c\"/>\n",
       "      <use x=\"63.720703\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"97.998047\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"145.800781\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"213.623047\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"262.939453\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"330.419922\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"378.222656\" xlink:href=\"#DejaVuSans-Bold-75\"/>\n",
       "      <use x=\"449.414062\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"498.730469\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"566.552734\" xlink:href=\"#DejaVuSans-Bold-54\"/>\n",
       "      <use x=\"621.515625\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"689.337891\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"738.654297\" xlink:href=\"#DejaVuSans-Bold-6d\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Gene -->\n",
       "     <g transform=\"translate(278.087706 314.307377)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 4781 347 \n",
       "Q 4331 128 3847 18 \n",
       "Q 3363 -91 2847 -91 \n",
       "Q 1681 -91 1000 561 \n",
       "Q 319 1213 319 2328 \n",
       "Q 319 3456 1012 4103 \n",
       "Q 1706 4750 2913 4750 \n",
       "Q 3378 4750 3804 4662 \n",
       "Q 4231 4575 4609 4403 \n",
       "L 4609 3438 \n",
       "Q 4219 3659 3833 3768 \n",
       "Q 3447 3878 3059 3878 \n",
       "Q 2341 3878 1952 3476 \n",
       "Q 1563 3075 1563 2328 \n",
       "Q 1563 1588 1938 1184 \n",
       "Q 2313 781 3003 781 \n",
       "Q 3191 781 3352 804 \n",
       "Q 3513 828 3641 878 \n",
       "L 3641 1784 \n",
       "L 2906 1784 \n",
       "L 2906 2591 \n",
       "L 4781 2591 \n",
       "L 4781 347 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-47\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 4056 2131 \n",
       "L 4056 0 \n",
       "L 2931 0 \n",
       "L 2931 347 \n",
       "L 2931 1631 \n",
       "Q 2931 2084 2911 2256 \n",
       "Q 2891 2428 2841 2509 \n",
       "Q 2775 2619 2662 2680 \n",
       "Q 2550 2741 2406 2741 \n",
       "Q 2056 2741 1856 2470 \n",
       "Q 1656 2200 1656 1722 \n",
       "L 1656 0 \n",
       "L 538 0 \n",
       "L 538 3500 \n",
       "L 1656 3500 \n",
       "L 1656 2988 \n",
       "Q 1909 3294 2193 3439 \n",
       "Q 2478 3584 2822 3584 \n",
       "Q 3428 3584 3742 3212 \n",
       "Q 4056 2841 4056 2131 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-6e\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-47\"/>\n",
       "      <use x=\"82.080078\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"149.902344\" xlink:href=\"#DejaVuSans-Bold-6e\"/>\n",
       "      <use x=\"221.09375\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Gwas -->\n",
       "     <g transform=\"translate(135.291976 254.742176)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 225 3500 \n",
       "L 1313 3500 \n",
       "L 1900 1088 \n",
       "L 2491 3500 \n",
       "L 3425 3500 \n",
       "L 4013 1113 \n",
       "L 4603 3500 \n",
       "L 5691 3500 \n",
       "L 4769 0 \n",
       "L 3547 0 \n",
       "L 2956 2406 \n",
       "L 2369 0 \n",
       "L 1147 0 \n",
       "L 225 3500 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-77\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-47\"/>\n",
       "      <use x=\"82.080078\" xlink:href=\"#DejaVuSans-Bold-77\"/>\n",
       "      <use x=\"174.462891\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"241.943359\" xlink:href=\"#DejaVuSans-Bold-73\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Efo -->\n",
       "     <g transform=\"translate(119.574838 437.228659)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 588 4666 \n",
       "L 3834 4666 \n",
       "L 3834 3756 \n",
       "L 1791 3756 \n",
       "L 1791 2888 \n",
       "L 3713 2888 \n",
       "L 3713 1978 \n",
       "L 1791 1978 \n",
       "L 1791 909 \n",
       "L 3903 909 \n",
       "L 3903 0 \n",
       "L 588 0 \n",
       "L 588 4666 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-45\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 2841 4863 \n",
       "L 2841 4128 \n",
       "L 2222 4128 \n",
       "Q 1984 4128 1890 4042 \n",
       "Q 1797 3956 1797 3744 \n",
       "L 1797 3500 \n",
       "L 2753 3500 \n",
       "L 2753 2700 \n",
       "L 1797 2700 \n",
       "L 1797 0 \n",
       "L 678 0 \n",
       "L 678 2700 \n",
       "L 122 2700 \n",
       "L 122 3500 \n",
       "L 678 3500 \n",
       "L 678 3744 \n",
       "Q 678 4316 997 4589 \n",
       "Q 1316 4863 1984 4863 \n",
       "L 2841 4863 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-66\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 2203 2784 \n",
       "Q 1831 2784 1636 2517 \n",
       "Q 1441 2250 1441 1747 \n",
       "Q 1441 1244 1636 976 \n",
       "Q 1831 709 2203 709 \n",
       "Q 2569 709 2762 976 \n",
       "Q 2956 1244 2956 1747 \n",
       "Q 2956 2250 2762 2517 \n",
       "Q 2569 2784 2203 2784 \n",
       "z\n",
       "M 2203 3584 \n",
       "Q 3106 3584 3614 3096 \n",
       "Q 4122 2609 4122 1747 \n",
       "Q 4122 884 3614 396 \n",
       "Q 3106 -91 2203 -91 \n",
       "Q 1297 -91 786 396 \n",
       "Q 275 884 275 1747 \n",
       "Q 275 2609 786 3096 \n",
       "Q 1297 3584 2203 3584 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-6f\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-45\"/>\n",
       "      <use x=\"68.310547\" xlink:href=\"#DejaVuSans-Bold-66\"/>\n",
       "      <use x=\"111.816406\" xlink:href=\"#DejaVuSans-Bold-6f\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Protein -->\n",
       "     <g transform=\"translate(401.429138 179.120911)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 588 4666 \n",
       "L 2584 4666 \n",
       "Q 3475 4666 3951 4270 \n",
       "Q 4428 3875 4428 3144 \n",
       "Q 4428 2409 3951 2014 \n",
       "Q 3475 1619 2584 1619 \n",
       "L 1791 1619 \n",
       "L 1791 0 \n",
       "L 588 0 \n",
       "L 588 4666 \n",
       "z\n",
       "M 1791 3794 \n",
       "L 1791 2491 \n",
       "L 2456 2491 \n",
       "Q 2806 2491 2997 2661 \n",
       "Q 3188 2831 3188 3144 \n",
       "Q 3188 3456 2997 3625 \n",
       "Q 2806 3794 2456 3794 \n",
       "L 1791 3794 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-50\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-50\"/>\n",
       "      <use x=\"73.291016\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"122.607422\" xlink:href=\"#DejaVuSans-Bold-6f\"/>\n",
       "      <use x=\"191.308594\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"239.111328\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"306.933594\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"341.210938\" xlink:href=\"#DejaVuSans-Bold-6e\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Variant -->\n",
       "     <g transform=\"translate(175.785442 342.002392)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 31 4666 \n",
       "L 1241 4666 \n",
       "L 2478 1222 \n",
       "L 3713 4666 \n",
       "L 4922 4666 \n",
       "L 3194 0 \n",
       "L 1759 0 \n",
       "L 31 4666 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-56\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-56\"/>\n",
       "      <use x=\"71.892578\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"139.373047\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"188.689453\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"222.966797\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"290.447266\" xlink:href=\"#DejaVuSans-Bold-6e\"/>\n",
       "      <use x=\"361.638672\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Tissue -->\n",
       "     <g transform=\"translate(352.644522 507.364118)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-54\"/>\n",
       "      <use x=\"68.212891\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"102.490234\" xlink:href=\"#DejaVuSans-Bold-73\"/>\n",
       "      <use x=\"162.011719\" xlink:href=\"#DejaVuSans-Bold-73\"/>\n",
       "      <use x=\"221.533203\" xlink:href=\"#DejaVuSans-Bold-75\"/>\n",
       "      <use x=\"292.724609\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Literature -->\n",
       "     <g transform=\"translate(27.294144 137.232366)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-4c\"/>\n",
       "      <use x=\"63.720703\" xlink:href=\"#DejaVuSans-Bold-69\"/>\n",
       "      <use x=\"97.998047\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"145.800781\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "      <use x=\"213.623047\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"262.939453\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"330.419922\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"378.222656\" xlink:href=\"#DejaVuSans-Bold-75\"/>\n",
       "      <use x=\"449.414062\" xlink:href=\"#DejaVuSans-Bold-72\"/>\n",
       "      <use x=\"498.730469\" xlink:href=\"#DejaVuSans-Bold-65\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#pa70d908b59)\">\n",
       "     <!-- Pathway -->\n",
       "     <g transform=\"translate(492.824606 77.601735)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path d=\"M 4056 2131 \n",
       "L 4056 0 \n",
       "L 2931 0 \n",
       "L 2931 347 \n",
       "L 2931 1625 \n",
       "Q 2931 2084 2911 2256 \n",
       "Q 2891 2428 2841 2509 \n",
       "Q 2775 2619 2662 2680 \n",
       "Q 2550 2741 2406 2741 \n",
       "Q 2056 2741 1856 2470 \n",
       "Q 1656 2200 1656 1722 \n",
       "L 1656 0 \n",
       "L 538 0 \n",
       "L 538 4863 \n",
       "L 1656 4863 \n",
       "L 1656 2988 \n",
       "Q 1909 3294 2193 3439 \n",
       "Q 2478 3584 2822 3584 \n",
       "Q 3428 3584 3742 3212 \n",
       "Q 4056 2841 4056 2131 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-68\" transform=\"scale(0.015625)\"/>\n",
       "       <path d=\"M 78 3500 \n",
       "L 1197 3500 \n",
       "L 2138 1125 \n",
       "L 2938 3500 \n",
       "L 4056 3500 \n",
       "L 2584 -331 \n",
       "Q 2363 -916 2067 -1148 \n",
       "Q 1772 -1381 1288 -1381 \n",
       "L 641 -1381 \n",
       "L 641 -647 \n",
       "L 991 -647 \n",
       "Q 1275 -647 1404 -556 \n",
       "Q 1534 -466 1606 -231 \n",
       "L 1638 -134 \n",
       "L 78 3500 \n",
       "z\n",
       "\" id=\"DejaVuSans-Bold-79\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-Bold-50\"/>\n",
       "      <use x=\"70.666016\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"138.146484\" xlink:href=\"#DejaVuSans-Bold-74\"/>\n",
       "      <use x=\"185.949219\" xlink:href=\"#DejaVuSans-Bold-68\"/>\n",
       "      <use x=\"257.140625\" xlink:href=\"#DejaVuSans-Bold-77\"/>\n",
       "      <use x=\"349.523438\" xlink:href=\"#DejaVuSans-Bold-61\"/>\n",
       "      <use x=\"413.878906\" xlink:href=\"#DejaVuSans-Bold-79\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pa70d908b59\">\n",
       "   <rect height=\"543.6\" width=\"558\" x=\"7.2\" y=\"7.2\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "graph = nx.from_pandas_edgelist(meta_rel_df, source=\"from_node\", target=\"to_node\")\n",
    "cmap = matplotlib.colors.ListedColormap([\"dodgerblue\", \"lightgray\", \"darkorange\"])\n",
    "meta_rel_df[\"from_node\"] = pd.Categorical(meta_rel_df[\"from_node\"])\n",
    "f = plt.figure(figsize=(10, 10))\n",
    "f.tight_layout()\n",
    "plt.subplot(1, 1, 1)\n",
    "nx.draw(\n",
    "    G=graph,\n",
    "    with_labels=True,\n",
    "    node_size=3000,\n",
    "    edgecolors=\"gray\",\n",
    "    node_color=\"skyblue\",\n",
    "    font_size=10,\n",
    "    font_weight=\"bold\",\n",
    "    width=0.75,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7651e0da-9863-4dab-9339-6efe57d5777e",
   "metadata": {},
   "source": [
    "### Reference\n",
    "\n",
    "* Duerr RH, Taylor KD, Brant SR, Rioux JD, Silverberg MS, Daly MJ, Steinhart AH, Abraham C, Regueiro M, Griffiths A, others. 2006. A genome-wide association study identifies IL23R as an inflammatory bowel disease gene. Science 314:1461–1463.\n",
    "\n",
    "* Finan C, Gaulton A, Kruger FA, Lumbers RT, Shah T, Engmann J, Galver L, Kelley R, Karlsson A, Santos R, others. 2017. The druggable genome and support for target identification and validation in drug development. Science translational medicine 9:eaag1166.\n",
    "\n",
    "* Momozawa Y, Mni M, Nakamura K, Coppieters W, Almer S, Amininejad L, Cleynen I, Colombel J-F, De Rijk P, Dewit O, others. 2011. Resequencing of positional candidates identifies low frequency IL23R coding variants protecting against inflammatory bowel disease. Nature genetics 43:43–47.\n",
    "\n",
    "* Zheng J, Brumpton BM, Bronson PG, Liu Y, Haycock P, Elsworth B, Haberland V, Baird D, Walker V, Robinson JW, John S, Prins B, Runz H, Nelson MR, Hurle M, Hemani G, Asvold BO, Butterworth A, Smith GD, Scott RA, Gaunt TR. 2019. Systematic Mendelian randomization and colocalization analyses of the plasma proteome and blood transcriptome to prioritize drug targets for complex disease."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
